internal/builder/builder.go

```diff
@@ -560,57 +567,13 @@ func (b *Builder) Save(logger logging.Logger, creatorMetadata CreatorMetadata) e
 
 func (b *Builder) addExplodedModules(kind string, logger logging.Logger, tmpDir string, image imgutil.Image, additionalModules []buildpack.BuildModule, layers dist.ModuleLayers) error {
 	collectionToAdd := map[string]toAdd{}
+	explodedAdditionalModules := explodeBuildModules(kind, tmpDir, additionalModules, logger)
 
-	type modInfo struct {
-		info     dist.ModuleInfo
-		layerTar string
-		diffID   v1.Hash
-		err      error
-	}
-
-	lids := make([]chan modInfo, len(additionalModules))
-	for i := range lids {
-		lids[i] = make(chan modInfo, 1)
-	}
-
-	for i, module := range additionalModules {
-		go func(i int, module buildpack.BuildModule) {
-			// create directory
-			modTmpDir := filepath.Join(tmpDir, fmt.Sprintf("%s-%s", kind, strconv.Itoa(i)))
-			if err := os.MkdirAll(modTmpDir, os.ModePerm); err != nil {
-				lids[i] <- modInfo{err: errors.Wrapf(err, "creating %s temp dir", kind)}
-			}
-
-			// create tar file
-			layerTar, err := buildpack.ToLayerTar(modTmpDir, module)
-			if err != nil {
-				lids[i] <- modInfo{err: err}
-			}
-
-			// generate diff id
-			diffID, err := dist.LayerDiffID(layerTar)
-			info := module.Descriptor().Info()
-			if err != nil {
-				lids[i] <- modInfo{err: errors.Wrapf(err,
-					"getting content hashes for %s %s",
-					kind,
-					style.Symbol(info.FullName()),
-				)}
-			}
-			lids[i] <- modInfo{
-				info:     info,
-				layerTar: layerTar,
-				diffID:   diffID,
-			}
-		}(i, module)
-	}
-
-	for i, module := range additionalModules {
```

Comment:
I move this logic to a new method `explodeBuildModules` to handle the parallelism and the logic to organize the BuildModules when they are flattened